gnuradio-runtime/lib/buffer_reader_sm.cc

```diff
@@ -0,0 +1,149 @@
+#ifdef BUFFER_DEBUG
+    std::ostringstream msg;
+    msg << "[" << d_buffer << ";" << this << "] "
+        << "items_available() WR_idx: " << d_buffer->d_write_index
+        << " -- WR items: " << d_buffer->nitems_written()
+        << " -- RD_idx: " << d_read_index << " -- RD items: " << nitems_read() << " (-"
+        << d_attr_delay << ") -- available: " << available;
+    GR_LOG_DEBUG(d_logger, msg.str());
+#endif
+
+    return available;
+}
+
+bool buffer_reader_sm::input_blkd_cb_ready(int items_required) const
+{
+    gr::thread::scoped_lock(*d_buffer->mutex());
+
+    return (((d_buffer->d_bufsize - d_read_index) < (uint32_t)items_required) &&
+            (d_buffer->d_write_index < d_read_index));
+}
+
+bool buffer_reader_sm::input_blocked_callback(int items_required, int items_avail)
+{
+    // Maybe adjust read pointers from min read index?
+    // This would mean that *all* readers must be > (passed) the write index
+    if (((d_buffer->d_bufsize - d_read_index) < (uint32_t)items_required) &&
+        (d_buffer->d_write_index < d_read_index)) {
+
+        // Update items available before going farther as it could be stale
+        items_avail = items_available();
+
+        // Find reader with the smallest read index that is greater than the
+        // write index
+        uint32_t min_reader_index = std::numeric_limits<uint32_t>::max();
+        uint32_t min_read_idx = std::numeric_limits<uint32_t>::max();
+        for (size_t idx = 0; idx < d_buffer->d_readers.size(); ++idx) {
+            if (d_buffer->d_readers[idx]->d_read_index > d_buffer->d_write_index) {
+                // Record index of reader with minimum read-index
+                if (d_buffer->d_readers[idx]->d_read_index < min_read_idx) {
+                    min_read_idx = d_buffer->d_readers[idx]->d_read_index;
+                    min_reader_index = idx;
+                }
+            }
+        }
+
+        // Note items_avail might be zero, that's okay.
+        items_avail += d_read_index - min_read_idx;
+        int gap = min_read_idx - d_buffer->d_write_index;
+        if (items_avail > gap) {
+            return false;
+        }
+
+#ifdef BUFFER_DEBUG
+        std::ostringstream msg;
+        msg << "[" << d_buffer << ";" << this << "] "
+            << "input_blocked_callback() WR_idx: " << d_buffer->d_write_index
+            << " -- WR items: " << d_buffer->nitems_written()
+            << " -- BUFSIZE: " << d_buffer->d_bufsize << " -- RD_idx: " << min_read_idx;
+        for (size_t idx = 0; idx < d_buffer->d_readers.size(); ++idx) {
+            if (idx != min_reader_index) {
+                msg << " -- OTHER_RDR: " << d_buffer->d_readers[idx]->d_read_index;
+            }
+        }
+
+        msg << " -- GAP: " << gap << " -- items_required: " << items_required
+            << " -- items_avail: " << items_avail;
+        GR_LOG_DEBUG(d_logger, msg.str());
+#endif
+
+        // Shift existing data down to make room for blocked data at end of buffer
```

Comment:
I'm wondering if a lot of this logic should be moved into the buffer class since it uses all the private variables